Vuln Table Improvements (CRASM-3557)

hawkishpolicy opened the pull request:
<!-- GitHub renders PRs such that soft line breaks are treated as hard
line breaks.  In order to make this template render as expected we
therefore have to avoid soft breaks and therefore will offend
markdownlint with long lines.  This is the reason for the markdownline
disable directive just below.

For more details see:
https://github.com/github/markup/issues/1050#issuecomment-294654762 -->
<!-- markdownlint-disable MD013 -->
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##
- refactored row and column definitions for "is_kev" and "is_kev_ransomware"
   - these key value pairs come in as booleans from the backend.
   - convert them to "N/A", "Yes", or "No" based on null, true, or false at vulnRow definition level.
      - this allows for proper display in the table, export as .csv, and sorting
   - string values are then converted back to booleans via the normalizeFilters function.
      - this allows for proper filtering.
      - appears previous boolean filtering was working becaause django orm treats "yes" or "no" as truthy and falsy.
- also moved component level filter logic into normalizeFilters function to deduplicate and improve readability.
- fixed issue where call was being made with undefined filters:
   - empty strings, null values, and or string containing only whitespace are normalized to undefined values via utility function cleanFilterModelItems.
   - previous filter values are cleared if field changes.
   - undefined filters are filtered out of filter model before making call via shouldTriggerFilterUpdate utility function.
   - handles edge case where filterModel is changing due to user edits by preventing premature requests
- replaced default filter badge with custom implementation:
   - it persists once a filter is applied and does not reset during filter changes.
   - only resets if filters are removed.
   - the default badge would reset during filter changes even though the table was still in a filtered state from the previous selection.
- reset filter model and fetch default data if user partially sets a filter model after filtering one, i.e. changes from "severity contains medium", selects "KEV" but does not select a value, then clicks outside the filter menu.
   - This prevents a mismatch between the current state of the table, filtered by severity medium, and the filter model.
   - Replicates similar functionality to clearing filters. 
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##
- Closes CRASM-3557
- Ensures boolean values of is_kev and is_kev_ransomware are used for filter.
- Ensures proper field names are used for sort
- Ensures string values of is_kev and is_kev_ransomware are used for display and export to .csv
- Other minor refinements and improvements.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##
- tested locally.
- unit tests.
<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [ ] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [ ] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.
- [ ] Bump major, minor, patch, pre-release, and/or build versions [as appropriate](https://semver.org/#semantic-versioning-specification-semver) via the `bump_version` script *if* this repository is versioned *and* the changes in this PR [warrant a version bump](https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).
- [ ] Create a pre-release (necessary if and only if the pre-release version was bumped).

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release (necessary if and only if the version was bumped).

hawkishpolicy committed:
Inital solution for sorting by kev and ransomware

jsalinasnttdata committed:
refs CRASM-3557: Fix sort for vulns

jsalinasnttdata committed:
refs CRASM-3557: Clean up I

jsalinasnttdata committed:
refs CRASM-3557: Undo mapping refactor

jsalinasnttdata committed:
refs CRASM-3557: Centralize sort field mapping in vulnerabilitiesTableUtils

- Move mapSortFieldToServerField from Vulnerabilities component to utils

jsalinasnttdata committed:
refs CRASM-3557: Map display fields to server fields for KEV/Ransomware sort and filter

- Maps is_kev_display -> is_kev and is_kev_ransomware_display -> is_kev_ransomware to fix backend communication while maintaining CSV export functionality.

hawkishpolicy committed:
Merge remote-tracking branch 'origin/Fix-Vuln-Table-Sort-CRASM-3557-jesse' into Fix-Vuln-Table-Sort-CRASM-3557

hawkishpolicy committed:
Update logic for is_kev and is_kev_ransomware:

- mock up of fix to simplify conversion of is_kev and is_kev_ransomware boolean value to string for display and .csv export.
  - boolean values are now converted to 'Yes' or 'No' strings for display and .csv export at the row level.
  - these "Yes" or "No" values are then converted back to boolean values for filtering via the normalizeFilters function in vulnerabilitiesTableUtils.ts.
  - this change ensures that the boolean values are correctly displayed and filtered in the vulnerabilities table.
  - prevents duplicate column definitions in the table.
  - prevents duplicate row definitions in the table.

hawkishpolicy committed:
Additional code cleanup

hawkishpolicy committed:
Add orgId to normalizeFilters and clean up code

hawkishpolicy committed:
Edit snapshot test:

- make snapshot test async so the tests waits state updates and mock api.
- delete old snapshot and create a new one.

hawkishpolicy committed:
Add logic to onFilterModelChange handler:

- Created new utility function shouldTriggerFilterUpdate
  - This function provides the following:
    - prevents the filter update if the filter model lacks values.
    - prevents the filter update when switching fields but not changing the value.
- Updated onFilterModelChange handler to use the new utility function

hawkishpolicy committed:
Add custom badge to filter menu:

- this badge indicates whether any filters are currently active
- implemented in CustomToolbar component
- integrated with Vulnerabilities page to track active filters
- old badge was based off of model state and would reset between filter changes even though the data was filtered.
  - set display to none

hawkishpolicy committed:
Update snapshots

hawkishpolicy committed:
Clear filter model on model change:

- sets value to undefine if new filter field does not match previous filter but item id is the same.
- This prevents stale filter values from persisting when changing filter fields.
- also cleans filter items with empty string values.

hawkishpolicy committed:
Minor improvements:

- setHasActiveFilters is now set based off a comparison of newKey generated from shouldTriggerFilterUpdate instead of the length of model.items. This ensures that the active filter state accurately reflects whether any filters with values are applied.
- trimmed white space from cleanedItems during normalization to avoid counting empty string filters as active.

hawkishpolicy committed:
add type check to .trim()

jsalinasnttdata commented on frontend/src/tests/pages/Vulnerabilities/vulnerabilitiesTable.test.tsx:
```diff
@@ -487,8 +487,25 @@ describe('Vulnerabilities component', () => {
+
+    await screen.findByRole('grid');
+
     expect(container.firstChild).toMatchSnapshot();
```
See pipeline error below, looks like we need to rebuild the snapshot.

hawkishpolicy replied:
Addressed via https://github.com/cisagov/XFD/pull/1396/commits/2fb6fdbfbb12af63bea775f009f12036220c970b

cduhn17 commented on frontend/src/pages/Vulnerabilities/Vulnerabilities.tsx:
```diff
@@ -442,7 +436,7 @@ export const Vulnerabilities: React.FC<VulnerabilitiesProps> = ({
         ),
         renderCell: (cellValues: GridRenderCellParams<VulnerabilityRow>) => {
-          const v = cellValues.row.is_kev_display;
+          const v = cellValues.row.is_kev;
```
Consider not using single letter variable name.

hawkishpolicy replied:
Addressed via [80d7ef3](https://github.com/cisagov/XFD/pull/1396/commits/80d7ef3561d839978b83c6066661e2e96516471e)

cduhn17 commented on frontend/src/utils/vulnerabilitiesTableUtils.ts:
```diff
@@ -140,5 +154,45 @@ export const normalizeFilters = (
+        item.value !== undefined && item.value !== null && item.value !== ''
+    )
+    .map((f) => ({
+      field: f.field,
```
Another example of single letter variable name. We would want to use descriptive naming.

hawkishpolicy replied:
Addressed via [80d7ef3](https://github.com/cisagov/XFD/pull/1396/commits/80d7ef3561d839978b83c6066661e2e96516471e)

cduhn17 commented on frontend/src/utils/vulnerabilitiesTableUtils.ts:
```diff
@@ -30,6 +30,18 @@ export const formatSeverity = (severity?: any) => {
   }
 };
 
+export const convertStringToBooleanValue = (field: string, value: any): any => {
```
What do we think of adding JSDoc comments to all functions? If there is value in having the comments we would need to make CRASM tickets to do it.

hawkishpolicy replied:
I'm not opposed.

hawkishpolicy replied:
Would be a larger conversation with the Frontend Team.

cduhn17 reviewed (commented):
See comments and you have checks failing. Easy fixes.

hawkishpolicy committed:
Clear filters & filter model on incomplete filters

- If user closes filter panel with incomplete filters, clear filter model and fetch data without filters.
- This prevents inconsistent states with partial filters applied.

hawkishpolicy committed:
Remove single letter variables

hawkishpolicy committed:
Refactor shouldTriggerFilterChange logic:

- removed Key based logic for detecting active filters
- directly check current filter items against previous model items.
- cleaned up unused refs related to filter state tracking

hawkishpolicy committed:
Clean up commented out code:

- vullnerabilitiesTable.test.tsx: Remove the commented out snapshot test.

hawkishpolicy committed:
Update unit test for Vulnerabilities table:

- Updated snapshot to reflect changes in filter handling.
- Added utility tests for filter model cleaning.
- Added utility tests for filter update triggering.

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Fix-Vuln-Table-Sort-CRASM-3557

hawkishpolicy committed:
Update snapshot after develop merge

hawkishpolicy committed:
Tiny change in test title to fix snapshot mismatch

hawkishpolicy committed:
Run tests with coverage

hawkishpolicy committed:
Update snapshot with correct CSS hash after fresh npm ci

hawkishpolicy commented:
> See comments and you have checks failing. Easy fixes.

Addressed.

cduhn17 reviewed (dismissed):
LGTM

jsalinasnttdata reviewed (dismissed):
LGTM

hawkishpolicy committed:
Merge branch 'develop' into Fix-Vuln-Table-Sort-CRASM-3557

hawkishpolicy committed:
Update snapshot for vulns table

hawkishpolicy committed:
Merge branch 'Fix-Vuln-Table-Sort-CRASM-3557' of https://github.com/cisagov/XFD into Fix-Vuln-Table-Sort-CRASM-3557

hawkishpolicy committed:
Merge remote-tracking branch 'origin/develop' into Fix-Vuln-Table-Sort-CRASM-3557

hawkishpolicy committed:
install testing-library/user-event

cduhn17 reviewed (approved):
LGTM - checks are now passing
